### DISCUSSION TITLE
Resolving based on Union type

### DISCUSSION THREAD

**Discussion by sm-the-explorer** [Netflix/dgs-framework#1125] [Q&A] [answered]
Hello, 

I am looking to resolve a field based on Union type. Want to check how I can do that ?

Example:-

Media is a **union of Book and Movie**. For suppose I am getting Book and Movie from different data sources and I would like to either one or both based the client request. If client is executing a query only for only Book type. I would like to resolve only for Book  type.

union Media = Book | Movie

Can you please guide me through if there any approach I can do this ?

Thank you

**Comment by springmonster** [comment, marked as answer]
Hi @sm-the-explorer , will these 3 links be helpful for you?

https://graphql.org/learn/schema/#union-types
https://netflix.github.io/dgs/advanced/type-resolvers-for-abstract-types/

This is an example of interface and union:
https://github.com/springmonster/netflix-dgs-example-java/tree/main/l-interfaceunion

Thanks!

**Reply by sm-the-explorer** [reply]
Thanks for the links @springmonster 

My question is not covered in the above links. To elaborate more I will refer to this example below.

https://github.com/springmonster/netflix-dgs-example-java/blob/main/l-interfaceunion/src/main/java/com/interfaceunion/datafetcher/MovieDataFetcher.java

    @DgsData(parentType = "Query", field = "movies")
    public List<Movie> movies() {
        return Lists.newArrayList(
                new ActionMovie("Crouching Tiger", 0),
                new ActionMovie("Black hawk down", 10),
                new ScaryMovie("American Horror Story", true, 10),
                new ScaryMovie("Love Death + Robots", false, 4)
            );
    }

The above  query can either return a ActionMovie or ScaryMovie and to get the list of ActionMovie's or ScaryMovie's imagine I need to call separate services. Now my question is about calling the service for ActionMovie or ScaryMovie only if the client requests in the query on corresponding type.

**Reply by springmonster** [reply]
Hi @sm-the-explorer , please check code below, will this be helpful? Thanks!

```
@DgsComponent
@Slf4j
public class MovieDataFetcher {

    @DgsData(parentType = "Query", field = "movies")
    public List<Movie> movies(DgsDataFetchingEnvironment environment) {
        List<Movie> movies = new ArrayList<>();

        environment.getMergedField().getSingleField().getSelectionSet().getSelections().forEach(selection -> {
            InlineFragment inlineFragment = (InlineFragment) selection;
            String name = inlineFragment.getTypeCondition().getName();
            if ("ActionMovie".equals(name)) {
                movies.addAll(fetchActionMovie());
            }
            if ("ScaryMovie".equals(name)) {
                movies.addAll(fetchScaryMovie());
            }
        });

        return movies;
    }

    private List<ActionMovie> fetchActionMovie() {
        log.info("fetch action movie!");
        return Lists.newArrayList(
                new ActionMovie("Crouching Tiger", 0),
                new ActionMovie("Black hawk down", 10)
        );
    }

    private List<ScaryMovie> fetchScaryMovie() {
        log.info("fetch scary movie!");
        return Lists.newArrayList(
                new ScaryMovie("American Horror Story", true, 10),
                new ScaryMovie("Love Death + Robots", false, 4)
        );
    }
}

**Reply by sm-the-explorer** [reply]
@springmonster  this worked. Thanks for the help :)
